### DISCUSSION TITLE
development toolbox does not work for Debian packages

### DISCUSSION THREAD

**Discussion by SirH2Os** [cockpit-project/cockpit#21275] [Q&A] [answered]
### Explain what happens

Hello :wave: 

I am trying to setup a development environment (using distrobox) and I've got some errors that do not let me work with, here is what I've done so far:

ps. following the documentation found [here](https://github.com/cockpit-project/cockpit/blob/main/HACKING.md)

1. Create the container
`distrobox create -i [ghcr.io/cockpit-project/tasks](http://ghcr.io/cockpit-project/tasks) -n cockpit-dev`

2. Enter the container
`distrobox enter cockpit-dev`

Where the problem begins:

Executing the tools/make-debs --quick command outputs the following:
`CONFIG   tmp/build-dist/Makefile
  GEN      cockpit-328.dev25+g132300e09/version.m4
  GEN      cockpit-328.dev25+g132300e09/tools/debian/copyright
  GEN      cockpit-328.dev25+g132300e09/configure
  GEN      cockpit-328.dev25+g132300e09/src/cockpit/_version.py
  DIST     cockpit-328.dev25+g132300e09.tar.xz
dpkg-buildpackage: info: source package cockpit
dpkg-buildpackage: info: source version 328.dev25+g132300e09-1
dpkg-buildpackage: info: source distribution UNRELEASED
dpkg-buildpackage: info: source changed by Cockpit <[cockpit@cockpit-project.org](mailto:cockpit@cockpit-project.org)>
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build .
dpkg-checkbuilddeps: error: Unmet build dependencies: debhelper-compat (= 13) dh-python gettext (>= 0.21) zlib1g-dev libkrb5-dev (>= 1.11) libxslt1-dev libglib2.0-dev (>= 2.50) libgnutls28-dev (>= 3.4.3) | gnutls-dev libsystemd-dev (>= 235) libpolkit-agent-1-dev libjson-glib-dev libpam0g-dev pkgconf systemd-dev | systemd (<< 253-2~) xsltproc xmlto docbook-xsl glib-networking python3 python3-pip python3-setuptools python3-wheel openssh-client procps python3-pytest-asyncio python3-pytest-timeout
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)`

And when I try to execute solely the make command, the following appears:
`make: *** No targets specified and no makefile found.  Stop.`

### Version of Cockpit

328

### Where is the problem in Cockpit?

Unknown or not applicable

### Server operating system

Ubuntu

### Server operating system version

22.04.1

### What browsers are you using?

Firefox, Brave

**Comment by martinpitt** [comment, marked as answer]
Hello @SirH2Os ! Our "tasks" development container is Fedora based. You cannot build Debian packages in it. You can work on the sources and build them the "upstream" way (./autogen.sh, make, etc., see https://github.com/cockpit-project/cockpit/blob/main/HACKING.md). If you want to build Debian packages, use a Debian tool like `schroot` or `pbuilder`, or your host system.

In our CI, we test Debian/Ubuntu by building them in a VM, e.g. `test/image-prepare ubuntu-2204` in your case.

**Comment by SirH2Os** [comment]
Hi @martinpitt, thanks for the advice! :smile: 

I've been trying to do what you have told me with a docker setup but I am struggling to find a way to solve the following issue:

I've gathered all the dependencies that ./autogen.sh requires and now while doing the make command the following error appears...

`=> ERROR [ 9/11] RUN cd cockpit; make                                                                                                                                                180.7s
------                                                                                                                                                                                       
 > [ 9/11] RUN cd cockpit; make:                                                                                                                                                             
0.553   GEN      src/ws/mock-dbus-tests.h                                                                                                                                                    
1.087   GEN      src/ws/mock-dbus-tests.c                                                                                                                                                    
1.674 make  all-am                                                                                                                                                                           
1.793 make[1]: Entering directory '/app/cockpit'                                                                                                                                             
2.365   CC       src/ws/cockpit_ws-main.o
4.913   CC       src/ws/libcockpit_ws_a-cockpithandlers.o
7.336   CC       src/ws/libcockpit_ws_a-cockpitauth.o
10.86   CC       src/ws/libcockpit_ws_a-cockpitcompat.o
12.46   CC       src/ws/libcockpit_ws_a-cockpitbranding.o
14.59   CC       src/ws/libcockpit_ws_a-cockpitchannelresponse.o
17.55   CC       src/ws/libcockpit_ws_a-cockpitchannelsocket.o
19.85   CC       src/ws/libcockpit_ws_a-cockpitcreds.o
22.40   CC       src/ws/libcockpit_ws_a-cockpitwebservice.o
26.32   AR       libcockpit-ws.a
26.36 ar: `u' modifier ignored since `D' is the default (see `U')
26.91   CC       src/common/libcockpit_common_a-cockpitchannel.o
30.54   CC       src/common/libcockpit_common_a-cockpitclosefrom.o
32.12   CC       src/common/libcockpit_common_a-cockpitcontrolmessages.o
34.68   CC       src/common/libcockpit_common_a-cockpiterror.o
36.36   CC       src/common/libcockpit_common_a-cockpitflow.o
38.27   CC       src/common/libcockpit_common_a-cockpithash.o
39.91   CC       src/common/libcockpit_common_a-cockpitjson.o
43.45   CC       src/common/libcockpit_common_a-cockpitlocale.o
45.84   CC       src/common/libcockpit_common_a-cockpitmemfdread.o
48.37   CC       src/common/libcockpit_common_a-cockpitpipe.o
52.60   CC       src/common/libcockpit_common_a-cockpitpipetransport.o
55.89   CC       src/common/libcockpit_common_a-cockpitsocket.o
58.31   CC       src/common/libcockpit_common_a-cockpitsystem.o
61.00   CC       src/common/libcockpit_common_a-cockpittemplate.o
63.64   CC       src/common/libcockpit_common_a-cockpittransport.o
66.72   CC       src/common/libcockpit_common_a-cockpitunicode.o
68.56   CC       src/common/libcockpit_common_a-cockpitwebfilter.o
70.86   CC       src/common/libcockpit_common_a-cockpitwebinject.o
73.01   CC       src/common/libcockpit_common_a-cockpitwebresponse.o
77.63   CC       src/common/libcockpit_common_a-cockpitwebserver.o
81.91   GEN      src/common/fail-html.c
82.52   CC       src/common/libcockpit_common_a-fail-html.o
83.36   AR       libcockpit-common.a
83.41 ar: `u' modifier ignored since `D' is the default (see `U')
83.74   CC       src/common/cockpitauthorize.o
85.65   CC       src/common/cockpitbase64.o
87.43   CC       src/common/cockpitconf.o
89.21   CC       src/common/cockpitfdpassing.o
90.68   CC       src/common/cockpitframe.o
92.44   CC       src/common/cockpithex.o
93.67   CC       src/common/cockpitjsonprint.o
95.18   CC       src/common/cockpitmemory.o
96.69   CC       src/common/cockpitwebcertificate.o
98.37   AR       libcockpit-common-nodeps.a
98.41 ar: `u' modifier ignored since `D' is the default (see `U')
99.03   CC       src/websocket/libwebsocket_a-websocket.o
102.3   CC       src/websocket/libwebsocket_a-websocketclient.o
105.2   CC       src/websocket/libwebsocket_a-websocketserver.o
108.2   CC       src/websocket/libwebsocket_a-websocketconnection.o
112.4   AR       libwebsocket.a
112.5 ar: `u' modifier ignored since `D' is the default (see `U')
112.8   CCLD     cockpit-ws
113.8   CC       src/common/cockpitclosefrom.o
115.5   CC       src/session/client-certificate.o
117.4   CC       src/session/session-utils.o
119.7   CC       src/session/session.o
122.4   CCLD     cockpit-session
123.6   CC       src/tls/main.o
125.4   CC       src/tls/certificate.o
127.1   CC       src/tls/client-certificate.o
128.9   CC       src/tls/connection.o
131.9   CC       src/tls/httpredirect.o
133.6   CC       src/tls/server.o
135.6   CC       src/tls/socket-io.o
137.4   AR       libcockpit-tls.a
137.5 ar: `u' modifier ignored since `D' is the default (see `U')
137.9   CCLD     cockpit-tls
138.6   CC       src/tls/cockpit-certificate-ensure.o
140.6   CCLD     cockpit-certificate-ensure
141.3   CC       src/tls/cockpit_wsinstance_factory-wsinstance-factory.o
143.1   CCLD     cockpit-wsinstance-factory
143.7   CC       src/ws/mock_pam_conv_mod_so-mock-pam-conv-mod.o
145.1   CCLD     mock-pam-conv-mod.so
146.0   CC       src/pam-ssh-add/pam_ssh_add_so-pam-ssh-add.o
148.1   CC       src/pam-ssh-add/libpam_ssh_add_a-pam-ssh-add.o
151.0   AR       libpam_ssh_add.a
151.0 ar: `u' modifier ignored since `D' is the default (see `U')
151.4   CCLD     pam_ssh_add.so
151.9   CC       src/ws/pam_cockpit_cert_so-pam_cockpit_cert.o
152.9   CCLD     pam_cockpit_cert.so
153.4   GEN      doc/man/cockpit.1
156.0   GEN      doc/man/cockpit-bridge.1
158.6   GEN      doc/man/cockpit-desktop.1
161.0   GEN      doc/man/cockpit-ws.8
163.7   GEN      doc/man/cockpit-tls.8
166.3   GEN      doc/man/cockpit.conf.5
168.6   GEN      doc/man/pam_ssh_add.8
171.0   GEN      src/ws/org.cockpit_project.cockpit.appdata.xml
180.6   GEN      src/appstream/org.cockpit_project.cockpit_sosreport.metainfo.xml
180.7 msgfmt: cannot locate ITS rules for src/appstream/org.cockpit_project.cockpit_sosreport.metainfo.xml.in
180.7 make[1]: *** [Makefile:5991: src/appstream/org.cockpit_project.cockpit_sosreport.metainfo.xml] Error 1
180.7 make[1]: Leaving directory '/app/cockpit'
180.7 make: *** [Makefile:1964: all] Error 2
------
Dockerfile:62
--------------------
  60 |     # --------------------------------------------------------------------------
  61 |     RUN cd cockpit; ./autogen.sh --prefix=/usr --enable-debug
  62 | >>> RUN cd cockpit; make
  63 |     RUN cd cockpit; ls -lah **
  64 |     RUN cd cockpit; make install
--------------------
ERROR: failed to solve: process "/bin/sh -c cd cockpit; make" did not complete successfully: exit code: 2

View build details: docker-desktop://dashboard/build/engine-multiarch/engine-multiarch0/r7do9nw4yp9mlzuxz9t5lv3xs`

**I am failing to understand why it is failing...**

**Heres my docker file**
`# --------------------------------------------------------------------------
# Base Image
# --------------------------------------------------------------------------
FROM debian:buster-slim AS build

# --------------------------------------------------------------------------
# Labels
# --------------------------------------------------------------------------
LABEL os=linux

# --------------------------------------------------------------------------
# Work Dir
# --------------------------------------------------------------------------
WORKDIR /app/

# --------------------------------------------------------------------------
# Install APT Necessary packages
# --------------------------------------------------------------------------
# install build & runtime dependencies

RUN apt update -y 
RUN apt install -y --no-install-recommends build-essential \
                                            libudev-dev \
                                            libssl-dev \
                                            gettext-base \ 
                                            dpkg-dev \ 
                                            gpg \ 
                                            git \
                                            autoconf \
                                            automake \
                                            ca-certificates \
                                            libglib2.0-dev \
                                            pkg-config \
                                            libsystemd-dev \
                                            libjson-glib-dev \
                                            libgnutls28-dev \
                                            libkrb5-dev \
                                            libpam0g-dev \
                                            systemd \
                                            gettext \
                                            xsltproc \
                                            xmlto \
                                            libappstream-glib-dev \
                                            sosreport

# --------------------------------------------------------------------------
# Clean OS
# --------------------------------------------------------------------------
RUN apt-get clean && apt-get autoclean && rm -rf /var/lib/apt/lists/*

# --------------------------------------------------------------------------
# Clone Repository
# --------------------------------------------------------------------------
RUN apt-get update && apt-get install -y ca-certificates

RUN git clone https://github.com/cockpit-project/cockpit.git

# --------------------------------------------------------------------------
# Build Cockpit
# --------------------------------------------------------------------------
RUN cd cockpit; ./autogen.sh --prefix=/usr --enable-debug
RUN cd cockpit; make
RUN cd cockpit; ls -lah **
RUN cd cockpit; make install`
